man/man1/ismir_query.1.ronn
ismir_query(1) -- A tool to look for an audio query file in an audio index.
=============================================

## SYNOPSIS
Usage: ismir_query [SETTINGS]

## DESCRIPTION
ismir_query takes an audio query file (a wav/mp3 file) and looks for it in the audio index.

If given a mp3 file, ismir will invoke ffmpeg to convert this file to 5512Hz WAVE format, before indexing.

If given a WAV file, you must ensure that it are in 5512Hz beforehand.


## OPTIONS
  -q [ --query ] arg           Audio query file

  -c [ --criteria ] arg        Criteria for hit; between 0 and 8192; default "2867.2" This is the max distance allowed for something to be called a hit.
  Anything above 3200 is very unlikely to be a correct result.

  -d [ --dbname ] arg          List of databases to query. At least one must be given

  -p [ --processed-query ] arg Preprocessed query file. Not currently implemented

## OUTPUT

The program will output a line per hit, as below:

   match in '/dr-dat/4/files/Batch33/Disc13/mp3_128kbps/P3_2200_0000_041202_001.mp3' at 00:01:35 with distance 1207

Distance is a number between 0 and 8192, indicating how "good" the hit is. The default criteria corresponds to 35% of 8192,
which seems to be a good value. Most hits will be with a distance in the 1200-1300 range.

## AUTHOR
* Jesper Sindahl Nielsen <jsn@statsbiblioteket.dk>
* Asger Askov Blekinge <abr@kb.dk>

changelog
scape-xcorrsound (2.1.0)
* ismir_query: Criteria input
* Formatting changes to manpages
* install ismir_query, ismir_merge, ismir_build_index binaries and manpages
* ismir: Do not fail on space in filenames
* Updated manpages with changes in CLI interfaces
* ismir_query: can search in multiple databases
* sound-match: uses boost param parsing. Takes critera as double now
* ismir: Do not output ffmpeg when working on mp3
* Ismir: can index mp3 files directly by ffmpeg converting on the fly
* Make/cmake part of setup
* Ismir_index: Map files are postfixed with map, not prefixed
* travis: try to install the nessesary packages through ubuntu, not gem
* travis: use the toolchain way to install ubuntu packages
* matrix eval of different versions of g++ and use exact ubuntu dist=trusty
* updates to include ismir documentation in ABOUT and man files
* Force install of g++ 4.8.
* Options for old cmake version.
* Tie to specific mustache version for ronn.
* Changed the silence detection threshold. Now more sound is required to declare it as 'non-silence'
* added how to compile
* Added c++11 as a compiler flag in cmake. Fixes #15.
* Faster query algorithm for sounds index based on some statistics.
* Added descriptions for algorithms in the Algorithms section in the ABOUT file.
* documentation updates
* toolspec updates
* Fixes #13.
* added scape toolspecs for xcorrSound waveform-compare
* Added --version to overlap-analysis and sound-match.
* Fixes #6
* travis does not depend on git
* travis includes the dependencies for Ronn

scape-xcorrsound (2.0.2)
?

scape-xcorrsound (2.0.0) trusty; urgency=low
  * Fixed the debian packaging
  * Scape micro site
  * Install Instructions and version refs updated
  * README updates to match SCAPE template - work in 
  progress
  * Added descriptions for installing on Mac OS X
  * Tool dependencies as rpm packages added
  * Applied patch to remove the tpie dependency
  * Add command line flag --pad-short-block to control the 
  padding of samples shorter than the block length with 
  silence. Also add diagnostics to show more detail of 
  correlation failure if --verbose is given.
  * Correlation of short waveforms and Cygwin build
  * Pad short blocks read from the end of samples with 
  silence, so that the correlation is used in the case of 
  samples shorter than the block length (rather than exiting 
  the loop, leaving minimumVal as 2 and indicating success).
  * Additional includes necessary for building on Cygwin.
  * [Travis] test .deb and binaries
  * Added debian package to travis
  * And corrected project language yet again.
  * added a similarity measure output on success
  * Removed cout stuff
  * fixed a bug in AudioStream
  * received updates for tpie
  * build index now supports piping
  * quick and dirty wavStdin reader
  * now also compiling with TPIE and avcodec
  * prepared fingerprint_db for 'hashing-style' queries, 
  changed to using 64 samples skips, and various other things
  * added minor progress info for index building
  * files for CMake to find libavcodec
  * Got the revindex for testing purposes. Written by 
  Mathias Rav
  * changing design to have several fingerprinting 
  strategies
  * integrationTests now works with other paths
  * Removed the old debian specs as cmake does that job now
  * insertion and query works
  * ismir_build_index now working
  * Add travis file
  * fast comparisons
  * Added more information on why a waveform comparison 
  failed
  * Wrote up man pages for the tools
  * Added options: --threshold and --channel along with the 
  appropriate modifications for the parameters
  * added number of seconds pr block as a parameter to 
  waveform-compare
  * Renamed xcorrSound tool to the more descriptive 
  'overlap-analysis'
  * lets use a relational database
  * changed some things up. Fingerprints are now proper 
  size fx.
  * minor corrections and another way of indexing audio
  * fixed long long. Now int64_t
  * added return codes to migrationQA.cpp
  * A tool to help in creating the fingerprinting function 
  for a data set
  * sound index stuff. Various additions and changes. next 
  step is to do some test runs and see if the idea works.
  * Made a better silence detector for migrationQA
  * Lots of changes. Next step is to normalize the sound to 
  get more correct matches
  * Initial go at building an audio index
  * Fixed AudioStream to be more wav compliant. Now accepts 
  wav files with extra parameters
  * implemented the fingerprinting function. Need to fiddle 
  with the magic constant to get good results
  * migrationQA: now with negative offset and better info
  * migrationQA: tool now works. Remember input order 
  matters\!
  * soundMatch now works in a streaming fashion
 -- Jesper Sindahl Nielsen <jsn@statsbiblioteket.dk>  Wed, 23 Apr 2014 16:48:08 +0200

scape-xcorrsound (1.0.1) trusty; urgency=low
  * script for making the xcorrSound tarball
  * cleaning up fftw
  * sound match is now working
  * Fixed bugs in cross correlation
 -- Jesper Sindahl Nielsen <jsn@statsbiblioteket.dk>  Mon, 14 Nov 2011 13:54:48 +0000

scape-xcorrsound (1.0.0) trusty; urgency=low
  * Initial version
 -- Jesper Sindahl Nielsen <jsn@statsbiblioteket.dk>  Mon, 07 Nov 2011 14:34:30 +0000
 


apps/ismir_query.cc
#include <boost/program_options.hpp>
#include <fingerprint_db.hh>
#include <iostream>
#include <string>
#include <vector>

using std::vector;

namespace po = boost::program_options;
namespace si = sound_index;

void query(std::string &dbname, std::string &queryFile, float criteria) {
    std::vector<std::string> res;
    si::fingerprint_db db;
    db.open(dbname);
    db.query_scan(queryFile, res, criteria);
    db.close();

    for (size_t i = 0; i < res.size(); ++i) {
        std::cout << res[i] << std::endl;
    }

}

template <typename T>
std::string to_string_with_precision(const T a_value, const int n = 6)
{
    std::ostringstream out;
    out.precision(n);
    out << std::fixed << a_value;
    return out.str();
}

void init(int argc, char *argv[]) {

    std::vector<std::string> dbname;
    std::string queryFile;
    std::string processedQueryFile;
    float criteria;
    
    po::options_description generic("Program options");
    generic.add_options()
        ("help,h", "Print help message and return, everything else is ignored");

    po::positional_options_description positional;
        positional.add("dbname", -1);

    double maxCriteria = si::fingerprint_db::macro_sz*sizeof(uint32_t)*8;
    double defaultCriteria = 0.35*maxCriteria;
    std::string criteriaHint = std::string("Criteria for hit; between 0 and ") +
                                to_string_with_precision(maxCriteria,1) +
                                std::string(". Default ") +
                                to_string_with_precision(defaultCriteria,1) +
                                std::string(". This is the max distance allowed for something to be called a hit");
    po::options_description hidden("Settings");
    hidden.add_options()
	("query,q", po::value<std::string>(), "Audio query file")
            ("criteria,c", po::value<float>(), criteriaHint.c_str())
	("dbname,d", po::value<std::vector<std::string>>(), "Database names")
	("processed-query,p", po::value<std::string>(), "Preprocessed query file");

    po::options_description all("Allowed options");
    all.add(generic).add(hidden);

    po::options_description visible("Program options");
    visible.add(generic);

    po::variables_map vm;
    po::store(po::command_line_parser(argc, argv).options(all).positional(positional).run(), vm);

    if (vm.count("help")) {
        std::cout << "Usage: ismir_query SETTINGS" << std::endl << std::endl;
        std::cout << generic << std::endl;

        std::cout << hidden << std::endl;

        exit(0);
    }

    bool queryExists = false;

    if (vm.count("query")) {
        queryFile = vm["query"].as<std::string>();
        queryExists = true;
    } else {
        std::cout << "A query file must be provided" << std::endl;
        exit(1);
    }

    if (vm.count("processed-query")) {
        processedQueryFile = vm["processed-query"].as<std::string>();
    } else {
        queryExists = true;
    }

      if (vm.count("criteria")) {
            criteria = vm["criteria"].as<float>();
        } else {
            criteria = defaultCriteria;
        }

    if (vm.count("dbname")) {
        dbname = vm["dbname"].as<std::vector<std::string>>();
    } else {
        std::cout << "dbname must be provided" << std::endl;
        exit(1);
    }

    for (size_t i = 0; i < dbname.size(); ++i) {
        std::string current_dbname(dbname[i]);

        query(current_dbname, queryFile, criteria);
    }

}

int main(int argc, char *argv[]) {
    init(argc, argv);
    return 0;
}
